File: PALIN.hs
-- Copyright (c) 2015 Linus Yang

module Main where
import Data.Char (ord, chr)
import qualified Data.Sequence as S
import Data.Foldable (toList)

process :: S.Seq Int -> Int -> Int -> S.Seq Int
process xs pl pr
  | pl < 0 = (S.<|) 1 (S.update (S.length xs - 1) 1 $ xs)
  | xl < 9 = let xl' = xl + 1 in S.update pl xl' . S.update pr xl' $ xs
  | otherwise = process (S.update pl 0 . S.update pr 0 $ xs) (pl - 1) (pr + 1)
  where xl = S.index xs pl

getPos :: Int -> (Int, Int)
getPos n | n `mod` 2 == 0 = (mid -1, mid)
         | otherwise = (mid, mid)
  where mid = n `div` 2

palinize :: [Int] -> [Int]
palinize xs = left ++ mid ++ right
  where len = length xs
        (left, right') = splitAt (len `div` 2) xs
        mid = if len `mod` 2 == 0 then [] else [head right']
        right = reverse left

greaterThan :: [Int] -> [Int] -> Bool
greaterThan [] [] = False
greaterThan (x:xs) (y:ys) | x > y = True
                          | x < y = False
                          | otherwise = greaterThan xs ys

needProcess :: [Int] -> Bool
needProcess xs = not $ greaterThan (reverse left) right
  where len = length xs
        (left, right') = splitAt (len `div` 2) xs
        right = if len `mod` 2 == 0 then right' else tail right'

next :: [Int] -> [Int]
next xs = if needProcess xs then
            palinize . toList $ process (S.fromList xs) pl pr
          else palinize xs
  where (pl, pr) = getPos $ length xs

printIntList :: [Int] -> IO ()
printIntList = mapM_ (\x -> putChar $ chr (x + ord '0'))

readIntList :: IO [Int]
readIntList = do
  xs <- getLine -- EOF will kill you if using getChar!
  return $ map (\x -> ord x - ord '0') xs

readLoop :: Int -> IO ()
readLoop 0 = return ()
readLoop n = do
  xs <- readIntList
  printIntList . next $ xs
  putChar '\n'
  readLoop (n - 1)

main :: IO ()
main = do
  inp <- getLine
  let n = read inp :: Int
  readLoop n

File: PALIN.cpp
// Copyright (c) 2015 Linus Yang

#include <string>
#include <iostream>

using namespace std;

int process(char *s, int pl, int pr)
{
    if (pl < 0) {
        return 1;
    } else if (s[pl] < '9') {
        s[pl] = s[pr] = s[pl] + 1;
        return 0;
    } else {
        s[pl] = s[pr] = '0';
        return process(s, pl - 1, pr + 1);
    }
}

void palinize_list(char *s, int n)
{
    int pl, pr;
    pr = n / 2;
    pl = n % 2 ? pr : pr - 1;
    while (pl >= 0 && pr < n) {
        if (s[pl] != s[pr]) {
            s[pr] = s[pl];
        }
        pl--;
        pr++;
    }
}

void next(string& str)
{
    int pl, pr;
    int ret = 0;

    int n = str.size();
    char *s = new char[n + 1];
    std::copy(str.begin(), str.end(), s);
    s[n] = '\0';

    pr = n / 2;
    pl = n % 2 ? pr : pr - 1;
    ret = process(s, pl, pr);
    if (ret) {
        s[n - 1] = '1';
        cout << 1;
    } else {
        palinize_list(s, n);
    }
    cout << s << endl;
    delete [] s;
}

bool can_palinize(string& str)
{
    int n = str.size();
    const char *s = str.c_str();

    int pl, pr;
    pr = n / 2;
    pl = n % 2 ? pr : pr - 1;

    while (pl >= 0 && pr < n) {
        if (s[pl] > s[pr]) {
            return true;
        } else if (s[pl] < s[pr]) {
            return false;
        }
        pl--;
        pr++;
    }
    return false;
}

string palinize(string &str)
{
    int n = str.size();
    char *s = new char[n + 1];
    std::copy(str.begin(), str.end(), s);
    s[n] = '\0';
    palinize_list(s, n);
    string result(s);
    delete [] s;
    return result;
}

int main(void)
{
    int i, n;
    string str;

    cin >> n;
    for (i = 0; i < n; i++) {
        cin >> str;
        if (can_palinize(str)) {
            cout << palinize(str) << endl;
        } else {
            next(str);
        }
    }
    return 0;
}
